src/components/calls-page/calls-page.tsx

```diff
@@ -70,6 +84,104 @@ export const CallsPage = () => {
     initialHotkey: "p",
   });
 
+  const callActionHandlers = useRef<Record<string, Record<string, () => void>>>(
+    {}
+  );
+  const callIndexMap = useRef<Record<number, string>>({});
+
+  useEffect(() => {
+    callIndexMap.current = {};
+    Object.keys(calls).forEach((callId, i) => {
+      callIndexMap.current[i + 1] = callId;
+    });
+  }, [calls]);
+
+  const muteToggleTimeoutRef = useRef<number | null>(null);
+
+  const handleToggleGlobalMute = () => {
+    if (muteToggleTimeoutRef.current !== null) return;
+
+    setIsMasterInputMuted((prev) => {
+      const newMuteState = !prev;
+
+      setTimeout(() => {
+        sendCallsStateUpdate();
+      }, 0);
+
+      return newMuteState;
+    });
+
+    setIsSettingGlobalMute(true);
+
+    muteToggleTimeoutRef.current = window.setTimeout(() => {
+      muteToggleTimeoutRef.current = null;
+    }, 300);
+
+    window.setTimeout(() => {
+      setIsSettingGlobalMute(false);
+    }, 1000);
+  };
+
+  const { connect, disconnect, isConnected } = useWebSocket({
+    onAction: (action, index) => {
+      if (action === "toggle_global_mute") {
+        handleToggleGlobalMute();
+        return;
+      }
+
+      if (typeof index !== "number") {
+        console.warn(
+          "Missing or invalid index for call-specific action:",
+          action
+        );
+        return;
+      }
+
+      const callId = callIndexMap.current[index];
+      if (!callId) {
+        console.warn("No callId found for index:", index);
+        return;
+      }
+
+      const handlers = callActionHandlers.current[callId];
+      if (!handlers) {
+        logger.yellow(`No handlers found for callId: ${callId}`);
+        return;
+      }
+
+      switch (action) {
+        case "toggle_input_mute":
+          handlers.toggleInputMute?.();
+          break;
+        case "toggle_output_mute":
+          handlers.toggleOutputMute?.();
+          break;
+        case "increase_volume":
+          handlers.increaseVolume?.();
+          break;
+        case "decrease_volume":
+          handlers.decreaseVolume?.();
+          break;
+        case "push_to_talk_start":
+          handlers.pushToTalkStart?.();
+          break;
+        case "push_to_talk_stop":
+          handlers.pushToTalkStop?.();
+          break;
+        default:
+          logger.yellow(`Unknown action: ${action}`);
+          break;
+      }
+    },
+    dispatch,
+    onConnected: () => {
+      sendCallsStateUpdate();
+    },
+    resetLastSentCallsState: () => {
+      resetLastSentCallsState();
+    },
+  });
+
```

Comment:
Nice code! But is it possible to break a big chunk of this out and put it in a hook?